### PR TITLE
Add build log clustering model deployment

### PR DESCRIPTION
**Pull request by Shreyanand**
## Related Issues and Dependencies
#305
#354

## This introduces a breaking change

- [ ] Yes
- [X] No

## This Pull Request implements

A notebook with sklearn pipeline to train and save a clustering model on build logs. It also tests a service deployed using a config file attached in this PR.

### CODE REVIEW

File: notebooks/data-sources/gcsweb-ci/build-logs/model_seldon.ipynb [hunk 1]
```diff
@@ -0,0 +1,562 @@
+{
```
Comment (oindrillac) [inline notebooks/data-sources/gcsweb-ci/build-logs/model_seldon.ipynb:1]: <p>can we get the credentials from the .env by default?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aicoe-aiops/ocp-ci-analysis/pull/371/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

Comment (oindrillac) [inline notebooks/data-sources/gcsweb-ci/build-logs/model_seldon.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #8.    # Check if your model exists on s3
</pre><p>Can we uncomment this line?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aicoe-aiops/ocp-ci-analysis/pull/371/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='13'/>

Comment (oindrillac) [reply notebooks/data-sources/gcsweb-ci/build-logs/model_seldon.ipynb:1]: it will be useful to check if the model is present on s3 as a sanity check

Comment (MichaelClifford) [reply notebooks/data-sources/gcsweb-ci/build-logs/model_seldon.ipynb:1]: <p>This should set the variables with <code>os.getenv()</code>  as ideally these are being stored as envs through the Kube secret. and then if not, via .env file. which would require a the use of <code>dotenv</code>   above. As well as an .env file in this dir specific to this task, right? </p>

Comment (MichaelClifford) [inline notebooks/data-sources/gcsweb-ci/build-logs/model_seldon.ipynb:1]: <p>Any chance we can hand label these 25 to provide some performance metrics here as well? Also, map the outputs to the correct human readable label "pass, infra failure, application failure" (I think)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aicoe-aiops/ocp-ci-analysis/pull/371/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='18'/>

File: notebooks/data-sources/gcsweb-ci/build-logs/model_seldon.ipynb [hunk 2]
```diff
@@ -0,0 +1,562 @@
...
+    "# objects"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "458a173c-ccb4-47b5-ab5f-4f3d666ae644",
+   "metadata": {},
+   "source": [
+    "# Test seldon deployment service \n",
+    "We use the deployment [config](seldon_deployment_config) to deploy a seldon service."
```
Comment (chauhankaranraj) [inline notebooks/data-sources/gcsweb-ci/build-logs/model_seldon.ipynb:429]:
```suggestion
    "We use the deployment [config](seldon_deployment_config.yaml) to deploy a seldon service."
```

**Review by chauhankaranraj — CHANGES_REQUESTED** [review]

File: notebooks/data-sources/gcsweb-ci/build-logs/model_seldon.ipynb (hunk 1 above)
Comment (chauhankaranraj) [reply notebooks/data-sources/gcsweb-ci/build-logs/model_seldon.ipynb:1]: <p>Maybe i'm being <em>extra</em> cautious but is it possible to sanity check that the results from the endpoint exactly match the results that we'd get if <code>test_list</code> was passed to <code>pipeline.predict</code> here in the notebook? Just to make sure seldon is using the correct model and its behaving as expected :)</p>

Comment (isabelizimm) [inline notebooks/data-sources/gcsweb-ci/build-logs/model_seldon.ipynb:1]: <p>is this the right order for these commented lines? seems like the <code>delete</code>  portion should go before the upload file, in case users just want to uncomment the whole block</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aicoe-aiops/ocp-ci-analysis/pull/371/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='13'/>

Comment (Shreyanand) [reply notebooks/data-sources/gcsweb-ci/build-logs/model_seldon.ipynb:1]: ++ added os.getenv()

Comment (Shreyanand) [reply notebooks/data-sources/gcsweb-ci/build-logs/model_seldon.ipynb:1]: @chauhankaranraj the test sets are now same.

@MichaelClifford good idea! I'll create an issue to create a test set and label it, I'll add Pep to it as well since he knows the data in and out.

Comment (MichaelClifford) [reply notebooks/data-sources/gcsweb-ci/build-logs/model_seldon.ipynb:1]: <p>On this note, in general I would prefer if control flow was not dictated by users needing to comment or un-comment sections of code manually. Is there any way you can check for the appropriate conditions instead?  </p>

Comment (Shreyanand) [reply notebooks/data-sources/gcsweb-ci/build-logs/model_seldon.ipynb:1]: Yes, I agree, the comment and un-comment flow is quite confusing. In this case, the commented out code deletes a file on s3. I could uncomment it and leave it like that but it seems a little risky since it's a delete operation. I could remove it all together since the upload_file operation would anyway update the same file. The latter seems more reasonable to me. wdyt?

Comment (chauhankaranraj) [reply notebooks/data-sources/gcsweb-ci/build-logs/model_seldon.ipynb:1]: > Is there any way you can check for the appropriate conditions instead?

Maybe something this would do the trick?
```python
ret = s3_resource.meta.client.list_objects_v2(Bucket=s3_bucket, Prefix='build-log-clustering/tfidf-kmeans/model.joblib"')
if ret['KeyCount'] > 0:
    # exists, dont upload again
else:
    # doesnt exist, go ahead with upload
```

But if the goal is to _always_ replace existing `model.joblib` with the current one, I think just doing `bucket.upload_file` without any checks might work, coz replaces existing object afaik :smile:

Comment (MichaelClifford) [reply notebooks/data-sources/gcsweb-ci/build-logs/model_seldon.ipynb:1]: yes, just remove the delete part

**Review by chauhankaranraj — APPROVED** [review]

**Comment by chauhankaranraj** [comment]
changes look good! /lgtm

**Review by isabelizimm — COMMENTED** [review]
/lgtm

**Comment by MichaelClifford** [comment]
:clap: :clap: :clap:

/approve
